File: src/data_rover/core/validation/validators/containment.py
from __future__ import annotations

from ..issue import Issue, Severity
from ..pipeline import EntityValidator
from ..scope import Scope


def _walk_reaches_cycle(indexes, start: str, safe: set[str]) -> bool:
    """Walk the first-parent chain from `start`; True if it reaches a cycle.

    `safe` memoizes nodes known to terminate, so a sweep over the whole model
    stays O(elements) overall instead of O(elements * depth).
    """
    seen: set[str] = set()
    node: str | None = start
    while node is not None and node not in seen and node not in safe:
        seen.add(node)
        node = indexes.first_parent(node)
    if node is None or node in safe:
        safe.update(seen)
        return False
    return True


class ContainmentValidator(EntityValidator):
    def validate_element(self, model, el) -> list[Issue]:
        # single-parent: each element contained at most once
        parents = model.indexes.parents_of(el.id)
        if len(parents) > 1:
            return [
                Issue(
                    Severity.ERROR,
                    f"Element {el.id} has {len(parents)} containment parents "
                    "(must have at most one)",
                    [el.id],
                )
            ]
        return []

    def validate_global(self, model, scope: Scope) -> list[Issue]:
        issues: list[Issue] = []
        indexes = model.indexes
        safe: set[str] = set()
        if scope.ids is None:
            # exhaustive sweep: report the first containment cycle reached from
            # any contained element (one representative issue per run, matching
            # the historical behaviour)
            for start in indexes.containment_parents:
                if _walk_reaches_cycle(indexes, start, safe):
                    issues.append(
                        Issue(
                            Severity.ERROR,
                            f"Containment cycle detected involving element {start}",
                            [start],
                        )
                    )
                    break
        else:
            # scoped run: the per-entity parent-chain walk covers exactly the
            # scoped elements (cycles entirely outside the scope are the full
            # sweep's responsibility)
            for entity_id in scope.ids:
                if entity_id not in model.elements:
                    continue
                if _walk_reaches_cycle(indexes, entity_id, safe):
                    issues.append(
                        Issue(
                            Severity.ERROR,
                            f"Containment cycle detected involving element {entity_id}",
                            [entity_id],
                        )
                    )
        return issues

File: engine/src/validation/validators/containment.ts
import type { ElementRec } from '../../model/records.ts';
import { errorIssue } from '../issue.ts';
import type { Run, Validator } from '../pipeline.ts';

/**
 * Whether the first-parent chain from `start` reaches a cycle. `safe` holds
 * elements known to reach a root, shared across one run; a chain that reaches
 * a cycle adds nothing to it.
 */
function reachesCycle(start: ElementRec, safe: Set<ElementRec>): boolean {
	const seen = new Set<ElementRec>();
	let node: ElementRec | null = start;
	while (node !== null && !seen.has(node) && !safe.has(node)) {
		seen.add(node);
		node = node.parents[0]?.source ?? null;
	}
	if (node === null || safe.has(node)) {
		for (const element of seen) safe.add(element);
		return false;
	}
	return true;
}

/**
 * An element has at most one containment parent. In the scope, every element
 * whose first-parent chain reaches a cycle is reported, the ones hanging
 * below the cycle included.
 */
export class Containment implements Validator {
	readonly checkName = 'containment';

	validateElement(run: Run, el: ElementRec): void {
		const n = el.parents.length;
		if (n > 1) {
			run.out.push(
				errorIssue(
					`Element ${el.id} has ${n} containment parents (must have at most one)`,
					[el.id],
					'structural'
				)
			);
		}
	}

	validateGlobal(run: Run, scope: readonly string[]): void {
		const safe = new Set<ElementRec>();
		for (const id of scope) {
			const el = run.model.findElement(id);
			if (el !== undefined && reachesCycle(el, safe)) {
				run.out.push(
					errorIssue(`Containment cycle detected involving element ${id}`, [id], 'structural')
				);
			}
		}
	}
}
